src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,389 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Internal.DeveloperExperience;
+
+namespace System
+{
+    internal unsafe class CrashInfo
+    {
+        /// <summary>
+        /// The kind or reason of crash for the triage JSON
+        /// </summary>
+        public enum CrashReason
+        {
+            Unknown = 0,
+            UnhandledException = 1,
+            EnvironmentFailFast = 2,
+            InternalFailFast = 3,
+        }
+
+        /// <summary>
+        /// Type of runtime (currently only NativeAOT is used)
+        /// </summary>
+        public enum RuntimeType
+        {
+            Unknown = 0,
+            Desktop = 1,
+            NetCore = 2,
+            SingleFile = 3,
+            NativeAOT = 4,
+        }
+
+        /// <summary>
+        /// This block is passed to the Watson DOTNET CLRMA provider and must not have any GC references
+        /// </summary>
+        private struct TriageBlockBuffer
+        {
+            public const int BufferSize = 8192;
+            public fixed byte Buffer[BufferSize];
+        }
+
+        private bool _comma;
+        private int _currentBufferIndex;
+        private int _reservedBuffer;
+        private static TriageBlockBuffer s_triageBuffer;
+
+        public CrashInfo()
+        {
+            _comma = false;
+            _currentBufferIndex = 0;
+            _reservedBuffer = 0;
+        }
+
+        /// <summary>
+        /// Writes the opening bracket and header to triage buffer
+        /// </summary>
+        /// <param name="reason"></param>
+        /// <param name="message"></param>
+        public void Open(RhFailFastReason reason, string message)
+        {
+            // Write the opening bracket and basic header which should never fail
+            bool success = OpenValue(default, '{');
+            Debug.Assert(success);
+            success = WriteHeader(reason, message);
+            Debug.Assert(success);
+        }
+
+        /// <summary>
+        /// Write the closing bracket. The triage buffer is ready.
+        /// </summary>
+        public void Close() => CloseValue('}');
+
+        /// <summary>
+        /// Write the exception information with fallbacks if the info doesn't fit in the fixed size triage buffer.
+        /// </summary>
+        /// <param name="exception">exception object</param>
+        public void WriteException(Exception exception)
+        {
+            ReadOnlySpan<char> key = "exception";
```

Comment:
Any reason not to lower this directly to the helper? Also, you can use the u8 string specifier and save yourself the encoding for this one.

Reply:
I'm not sure what you mean.   This key is passed 3 times so creating a local seems better.